### DISCUSSION TITLE
Documentation "Bug"

### DISCUSSION THREAD

**Discussion by trent-shipley** [sqlalchemy/sqlalchemy#8821] [Usage Questions] [answered]
At:

[https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#special-modifiers-within-group-filter](https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#special-modifiers-within-group-filter)

The fist code example is not formatted correctly.  Specifically, in the first unnest and percentile example the SQL produced by the input Python code is not differentiated from the example Python code.  See attached.

This caused me momentary confusion while trying to learn the fundamentals of SQLalchemy.

![chrome_sqlalchemy_malformed_tutorial_2022-11-17](https://user-images.githubusercontent.com/13461182/202481628-54332ff1-cb62-4d50-803a-38e0eead8fdc.png)

**Comment by CaselIT** [comment]
Hi,

You mean with the yellow box like in the second snippet?

**Comment by trent-shipley** [comment]
Exactly.  And the same little error occurs frequently thereafter on the
page.  There were probably five or six or the same thing further down the
page.v

Trent

**Comment by zzzeek** [comment]
that's a print statement, and the output is what you'd see in your python console.    The second example, where there's a yellow box, that's SQL logging output.   the yellow boxes only surround SQL as it's being executed to the database.

When you see an example like this:


```py
>>> print('hi')
hi
```

that's just a print.  See https://docs.python.org/3/library/doctest.html for background on what this format is.

**Reply by zzzeek** [reply]
you can see this in the second example, where the yellow boxes surround the SQL logging output, but the result of the print statement is in between them, not within a yellow box.

**Reply by trent-shipley** [reply]
zi nee, So since the output is directly from a simple print statement and should appear on the console.

So in the given example the output seems to be partially parsed with some Python and SQL elements.

Not being able to tell the difference between having the Core and OPM verbosely emit parsed SQL and what would be sent to terminal standard out my be caused by the mechanics of working in Jupyter.

How does the instance you say correctly shows the result of a print statement sent to the terminal differ from this example which seems to emit parsed SQL from a simple print statement.

![chrome_lateral corellation_simple_print](https://user-images.githubusercontent.com/13461182/202504327-cebd963e-221a-4369-ba5c-30c0fe1a838a.png)

**Reply by CaselIT** [reply]
that seems a bug

**Reply by trent-shipley** [reply]
The very bottom of the page shows two highlighting options. 
There seems to be:

- [ ] A function returns a calculated value : : these are consistently presented as standard out to terminal *not* highlighted as SQL (not shown).
- [ ] with ... print(results.all()) : : These are always highlighted as SQL output _per se_.  See the first example in the included image.
- [ ] There are simple print(the Core or ORM statement directly) : : These always have some parsing and may print as pure SQL. These are sometimes highlighted as SQL and sometimes not.  See the SQL CASE expression or fragment in the included image. In the last example the print statement is printing the explicitly compiled complete SQL query, but the complete (not fragmentary) output SQL statement is not highlighted.  

(Also elsewhere in the document Core and ORM statements are consistently assigned to the variable 'stmt', but in the very last example on this page the variable 's' is used for the same purpose, which is distracting.)

![chrome_bottom_of_page](https://user-images.githubusercontent.com/13461182/202554945-7c333dc1-eca4-43d1-aa60-49d7327d1823.png)

**Reply by zzzeek** [reply]
indeed that LATERAL example is inconsistent with the practices of the previous example.

so we would determine that one of them needs to be changed (with the requisite search and replace throughout the rest of the docs).

while at the same time, if yellow boxes are put around printed SQL, that does not imply yellow boxes are put for *all* print statements.   just the ones that generate SQL.

**Comment by trent-shipley** [comment]
"just the ones that generate SQL."

I don't think this is well defined.


   - print(results.all()) is always considered to generate SQL (+
   retrieve results)
   - print(Core or ORM statement) seems to parse the statement to SQL and
   print the parsed results.
      - All the func.function().type examples return a SQL, not Python
      type, but are not treated as returning SQL.  -
      - When SQL expressions which are not full statements are returned
      they are treated as not SQL for some reason.
      - And when a simple print statement prints what is undeniable a
      full-fledged SQL statement it may or may not be highlighted as SQL.

The determining factor seems to be *how* the output gets parsed and
printed, not *what* is parsed and emitted.  Almost all of the examples
arguably print whole or fragmentary SQL.  It's kind of confusing to me as a
reader.  It makes me think about something I shouldn't even notice.  I'd be
happier with a clear and consistent separation of code from output, or
maybe three categories, code, parsed Core and ORM, and output.  but code,
parsed Core and ORM considered to be sufficiently SQL, and other parsed
Core and ORM plus all other output is ... well it's hard, and it should be
not merely irrelevant, but *actually helpful.*

**Reply by CaselIT** [reply]
>    - print(results.all()) is always considered to generate SQL (+
>    retrieve results)

the sql is generated by the execute, not the print. That only outputs the result list.

In all other cases if there is an execute then there should be a "highlited as sql" section.
if there is no execute then a normal print should be there.

There are definitely section where that's not true, like the one you pointed above.

> The determining factor seems to be *how* the output gets parsed and
> printed, not *what* is parsed and emitted

the difference here is the sql is being emitted to the database or only printed to standard output.
if it's emitted to db it gets the yellow box, if it's printed it gets no box

**Comment by trent-shipley** [comment]
Item 1:

OK.  If the results of an "execute" should get highlighted but results of a
simple print() shouldn't because that' isSQLalchemy team wants, then that
is not in the Tutorial consistently, and should be fixed.

Item 2:

As a reader with intermediate experience with SQL dialects, but only one
Python class and even less experience with ORMs ,how is the above
distinction going to help me learn SQLalchemy from the Tutorial?

As a learner typing in the examples in the Tutorial I'm interested in the
parsed SQL and how it relates to the Core and ORM languages, whether it
comes from the printed "results" of an execution or from the print function
printing the backend __str__ function of SQLalchemy which parses the Core
or ORM code behind the scenes.  So I would argue a three way distinction
would do me more good.  1) input code,  2) parsed SQL, no matter where
it came from, and 3) returned results which return like from a function
whether from an SQLalchemy execute or other Python function.  (It might be
pedagogically helpful to distinguish between results returned from a
SQLalchemy execution and running functions, which would mean there are four
categories.)

The above paragraph is an argument for a Tutorial feature refactoring or
re-engineering ... later.  I'm not sure how the distinction you want to
make is optimal for me as a learner.

Sincerly,

Trent

**Reply by zzzeek** [reply]
I think we should have a box, or at least highlighting, around *all* SQL with some kind of non-intrusive distinction if this SQL is generated from logging or from a print statement.    

what do we think of that idea?

**Reply by CaselIT** [reply]
that's what I was proposing in the changeset. I've left a couple of question there, can you take a look? https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4349

**Comment by CaselIT** [comment, marked as answer]
I've done some changes at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4349 but I don't think it's an improvement removing the `{opensql}` from the prints. We may still want to format the sql, but maybe in a different way from what's used by `{opensql}` that's kept for executions

**Reply by CaselIT** [reply]
the documentation for v2 should be all or almost all been updated to differentiate print sql strings and executed sql queries
https://docs.sqlalchemy.org/en/20/

**Comment by jvanasco** [comment]
It looks like there is the potential for 3 usage patterns being implemented:

* {printsql} - @CaselIT's idea for handling `print()` statements.
* {logsql} - my suggestion for what would be emitted to the database behind the scenes, i,e. logged via the echo option.
* {opensql} - or another command, for showing raw sql when needed.
